### ISSUE TITLE
Plans for Open Container Format compatibility?

### ISSUE THREAD

**Issue by webmaven** [tailhook/vagga#41]
Now that [opencontainers.org](http://www.opencontainers.org/) has launched, and the [runC](https://github.com/opencontainers/runc) implementation in Go is released, are there any plans to make Vagga compatible with these?

**Comment by joehillen** [comment]
Someone would need to write an runC implementation in Rust first.

**Comment by webmaven** [comment]
Given that runC is basically a copy of appC, don't you have a substantial
franction of that in
https://github.com/tailhook/vagga/tree/master/src/container already?

**Comment by tailhook** [comment]
Sure, I plan to do that (sorry, development is a bit slow now, because I should do rust 1.0 porting, which is a huge and atomic piece of work).

But I don't think you should look forward to it too much. Because vagga has two powerful things:

1. It's a fully unprivileged container runner (runc doesn't able to run unprivileged AFAIK). And there are limitation of what can be run unprivileged. So vagga will have somewhat limited features.
2. Vagga is a powerful and easy to use build tool for containers. Opencontainer doesn't say anything about how containers would be built.

So basically integration looks like a way to import OCP container and to export to OCP container. Which you can basically do now (by using `!Tar` command and by tar'ing `.vagga/container_name` folder respectively), except some vagga.yaml <-> config.json convertation may help.

Note also that as part of No.1 above, vagga not only runs unprivileged, it also does its best to isolate application to never touch any files outside of the project dir. So it will likely never be able to mount arbitrary directories and support arbitrary privileges.

Anyway having rusty libcontainer and simple runc-like utility would be nice. I even have to consider publish it earlier than vagga for rust 1.x, because it's smaller in scope. But I'm not sure I'm able to do good API (and I'm not sure libcontainer's API is rich enough for vagga itself, if I would copy that). Sure `src/container/container.rs` is mostly same thing, except it was created specifically for vagga.

Thoughts?

**Comment by webmaven** [comment]
Well, I was looking for a copy of runC in Rust (mostly to see if I should consider writing such myself as a project to learn Rust) when I came across vagga.

So I figured I'd ask if the equivalent work was already planned and I shouldn't bother duplicating effort.

It sounds as though vagga is different enough from Docker that such a library would be of very limited utility to it, but that a portion of vagga's existing code could be forked off to start writing an actual runC implementation.

**Comment by webmaven** [comment]
So, @tailhook, is a compatible version of runC.rs something you want to work on yourself, or not?

**Comment by tailhook** [comment]
Well, I'm more interested in container library that may be used for both runc and vagga. But not in the runc binary itself. But I may find time for it only after I'll port vagga to rust 1.x.

**Comment by tailhook** [comment]
@webmaven, I've started some work at https://github.com/tailhook/unshare, its low-level library for dealing with namespaces/chroots, etc. It's written from scratch, so it's mostly stub now, but should be done shortly, as I have already done that twice (This time I want to get rid of C code so that's why I'm rewriting)

After it will be in working stage we also need:
* dealing with mountpoints
* handling signals (probably on top of eventfd/mio?)
* setting up network (is it a part of runc for now?)

Then we can build runc on top of all this stuff.

Bonus points would be:
* Some super-minimal musl-linked in-container process supervisor (i.e. init), it's not in current runc, too, right? Because it's wrong to run arbitrary application as pid 1. 
* Cgroup handling library (it's not clear whether they are part of runc too)  

Thoughts?

**Comment by tailhook** [comment]
Okay this issue doesn't represent current status any more.

We have pretty good `unshare` crate, handling namespaces stuff. The `signal` crate that works with signals. And just recently started `libmount` crate which handles mount points (not everything from vagga moved there yet).

With the time passed it looks like we don't need any runc compatibility. At least nobody complained on it since then.

While my opinion may be a little bit biased (because I'm the author of vagga), but I think vagga provides higher level building tools and versioning of containers, with a user friendly UI, rather than pure runner. For anybody not familiar with vagga reading this, I would describe it like: vagga is the tool that "gets base system, installs package X with version Y, and then is able to run command in that container or upload the image to a repository". On the other hand, runc is of the kind "download that image and run something in it". It looks like they are on the opposite ends of the spectrum.

We could also allow to download runc images in vagga. But again (in my biased opinion) it looks like vagga makes building images easier than runc and docker, so nobody needs it. So unless someone does a pull request, it doesn't going to happen.

Anyway, it would probably be nice if someone would write a runc clone with `unshare` and `libmount` so we could reuse most of the mission-critical code.

**Comment by webmaven** [comment]
OK, so I am (tentatively) still interested in writing a runC clone in Rust (probably using unshare and libmount). I'll ping you if/when I get the project started.
